src/persist-txn/src/txn_cache.rs

```diff
@@ -590,86 +565,85 @@ impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64> TxnsCacheState
-            // timestamps if none of the writes need to be compacted, so that
-            // compaction isn't `O(compact calls * data shards)`.
-            let data_id = match self.datas_min_write_ts.peek() {
-                Some(Reverse((ts, _))) if ts < &self.since_ts => {
-                    let Reverse((_, data_id)) = self.datas_min_write_ts.pop().expect("just peeked");
-                    data_id
-                }
-                Some(_) | None => break,
-            };
-            let times = self
-                .datas
-                .get_mut(&data_id)
-                .expect("datas_min_write_ts should be an index into datas");
+    /// Informs the cache that all registers and forgets less than ts have been
+    /// applied.
+    pub(crate) fn mark_register_applied(&mut self, ts: &T) {
+        self.unapplied_registers
+            .retain(|(_, register_ts)| ts < register_ts);
+        debug_assert_eq!(self.validate(), Ok(()));
+    }
+
+    /// Compact the internal representation for `data_id` by removing all data
+    /// that is not needed to maintain the following invariants:
+    ///
+    ///   - The latest write and registration for each shard are kept in
+    ///     `self.datas`.
+    ///   - All unapplied writes and registrations are kept in `self.datas`.
+    ///   - All writes in `self.datas` are contained by some registration in
+    ///     `self.datas`.
+    fn compact_data_times(&mut self, data_id: &ShardId) {
+        let Some(times) = self.datas.get_mut(data_id) else {
+            return;
+        };
+
+        debug!("cache compact {:.9} times={:?}", data_id.to_string(), times);
+
+        if let Some(unapplied_write_ts) = self
+            .unapplied_batches
+            .first_key_value()
+            .map(|(_, (_, _, ts))| ts)
+        {
             debug!(
-                "cache compact {:.9} since={:?} times={:?}",
+                "cache compact {:.9} unapplied_write_ts={:?}",
                 data_id.to_string(),
-                self.since_ts,
-                times
+                unapplied_write_ts,
             );
-
-            // Pop registrations that are forgotten before since_ts. We can
-            // stop as soon as something is >= or not forgotten.
-            while let Some(x) = times.registered.front() {
-                if let Some(forget_ts) = x.forget_ts.as_ref() {
-                    if forget_ts < &self.since_ts {
-                        times.registered.pop_front();
-                        continue;
-                    }
-                }
-                break;
-            }
-
-            // Pop any write times before since_ts. We can stop as soon
-            // as something is >=.
             while let Some(write_ts) = times.writes.front() {
-                if write_ts < &self.since_ts {
-                    times.writes.pop_front();
-                } else {
+                if times.writes.len() <= 1 || write_ts >= unapplied_write_ts {
```

Comment:
nit: Isn't `times.writes.len() <= 1` redundant with `let Some(write_ts) = times.writes.front()`?

Reply:
True, I could replace that with `times.writes.len() == 1`.

Reply:
Ah, oh no biggie on that. I was just reading it incorrectly as making sure the pop didn't panic, but it's actually what's protecting that we keep the latest write. Nevermind!